File: LiteDb.OneBella/Views/MainWindow.axaml.fs
namespace OneBella.Views

open System
open Avalonia
open Avalonia.Controls
open Avalonia.Controls.ApplicationLifetimes
open Avalonia.Interactivity
open Avalonia.Markup.Xaml
open LiteDb.Studio.Avalonia.Repo
open LiteDb.Studio.Avalonia.UseCases
open OneBella
open OneBella.UseCases
open OneBella.ViewModels
open LiteDB
open Avalonia.Input
open System.Windows.Input


type MainWindow () as this = 
    inherit Window ()

    let mutable fly = Unchecked.defaultof<Flyout>
    do this.InitializeComponent()
    do this.Opened |> Observable.add (fun _ -> this.OpenConnectionWindowCLick(this, null))
    do this.Closed |> Observable.add (fun _ ->
         let vm = this.DataContext :?> MainWindowViewModel
         Repo.Disconnect()
         for d in vm.DbItems do
            if d.IsConnected then
               d.Disconnect()
         )
    do this.KeyDown |> Observable.add (fun arg ->
        if (arg.Key = Key.F5) then
            let vm = this.DataContext :?> MainWindowViewModel
            if not (vm.SelectedTab = Unchecked.defaultof<ScriptViewModel> ) then
                let t = vm.SelectedTab.RunCommand :> ICommand
                t.Execute(null)
                )

    member private x.ScriptTabFlyoutOpened(sender:obj, _:EventArgs) =
        fly <- sender :?> Flyout

    member private x.ScriptTabFlyoutClickYes(sender:obj, _:RoutedEventArgs) =
        let main = x.DataContext :?> MainWindowViewModel
        if (main.Tabs.Count > 1)then
            use tab = sender :?> Button |> (fun b -> b.DataContext :?> ScriptViewModel)
            ignore(main.Tabs.Remove tab)
        fly.Hide()

    member private x.ScriptTabFlyoutClickNo(_:obj, _:RoutedEventArgs) =
        fly.Hide()

    member private x.OpenConnectionWindowCLick(_:obj, _:RoutedEventArgs)=
        let rec showAddWindow mainWindow conVm=
            async {
            let w = AddConnectionWindow(conVm)
            do! w.ShowDialog(mainWindow) |> Async.AwaitTask
            let! con = w.SelectFileTask |> Async.AwaitTask
            let vm = x.DataContext :?> MainWindowViewModel
            try
               conVm.Error <- "" 
               vm.Connect(con)
               StoredConnUseCase.Save(con, StoredConnUseCase.Create(fun _ -> Repo.GetDb()))

            with
            | exc -> 
              let err = exc.Message
              conVm.Error <- err
              do! showAddWindow mainWindow conVm
            }
                

        let run() =
            async {
            match Application.Current.ApplicationLifetime with
            | :? IClassicDesktopStyleApplicationLifetime as desktop ->
                //let savedConnections = Repo.getConnSettings()
                let uc =StoredConnUseCase.Create(fun _ -> Repo.GetDb())
                let p = StoredConnUseCase.LoadAll uc |> Seq.toArray
                let vm =ConnectionViewModel(p)
                do! showAddWindow desktop.MainWindow vm
            | _ -> ()
        }
        run() |> Async.StartImmediateAsTask |> ignore


    member private this.InitializeComponent() =
#if DEBUG
        this.AttachDevTools()
#endif
        AvaloniaXamlLoader.Load(this)

File: LiteDb.Studio.Avalonia/Views/MainWindow.axaml.cs
using System.Windows.Input;
using Avalonia;
using Avalonia.Controls;
using Avalonia.Controls.ApplicationLifetimes;
using Avalonia.Input;
using Avalonia.Interactivity;
using Avalonia.Markup.Xaml;
using LiteDb.Studio.Avalonia.UseCases;
using LiteDb.Studio.Avalonia.ViewModels;

namespace LiteDb.Studio.Avalonia.Views;

public partial class MainWindow : Window
{
    private Flyout? fly;

    private MainWindowViewModel _vm => (this.DataContext as MainWindowViewModel)!;
    
    public MainWindow()
    {
        InitializeComponent();

        this.Opened += (_, __) => OpenConnectionWindowClick(this, null);

        this.Closed += (_, __) =>
        {
            var vm = _vm;
            Repo.Repo.Disconnect();
            foreach (var d in vm.DbItems)
            {
                if (d.IsConnected)
                    d.Disconnect();
            }
        };

        this.KeyDown += (sender, e) =>
        {
            if (e.Key == Key.F5)
            {
                var vm = _vm;
                if (vm.SelectedTab != null)
                {
                    ICommand cmd = vm.SelectedTab.RunCommand;
                    cmd.Execute(null);
                }
            }
        };
    }

    private void ScriptTabFlyoutOpened(object sender, EventArgs e)
    {
        fly = (Flyout)sender;
    }

    private void ScriptTabFlyoutClickYes(object sender, RoutedEventArgs e)
    {
        var main = _vm;
        if (main.Tabs.Count > 1)
        {
            var button = (Button)sender;
            var tab = (button.DataContext as ScriptViewModel)!;
            main.Tabs.Remove(tab);
        }

        fly?.Hide();
    }

    private void ScriptTabFlyoutClickNo(object sender, RoutedEventArgs e)
    {
        fly?.Hide();
    }

    private void OpenConnectionWindowClick(object sender, RoutedEventArgs? e)
    {
        async Task ShowAddWindowAsync(Window mainWindow, ConnectionViewModel conVm)
        {
            var w = new AddConnectionWindow(conVm);

            await w.ShowDialog(mainWindow);
            var con = await w.SelectFileTask;
            
            try
            {
                conVm.Error = "";
                _vm.Connect(con);
                var uc = StoredConnUseCase.Create(Repo.Repo.GetDb);
                StoredConnUseCase.Save(con, uc);
            }
            catch (Exception exc)
            {
                conVm.Error = exc.Message;
                await ShowAddWindowAsync(mainWindow, conVm);
            }
        }

        async Task RunAsync()
        {
            if (Application.Current?.ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop)
            {
                var uc = StoredConnUseCase.Create(Repo.Repo.GetDb);
                var savedConnections = StoredConnUseCase.LoadAll(uc).ToArray();
                var vm = new ConnectionViewModel(savedConnections);
                await ShowAddWindowAsync(desktop.MainWindow!, vm);
            }
        }

        RunAsync().ConfigureAwait(false);
    }

    private void InitializeComponent()
    {
#if DEBUG
        this.AttachDevTools();
#endif
        AvaloniaXamlLoader.Load(this);
    }
}